docs/v1-proposed-api.md

```diff
@@ -0,0 +1,280 @@
+ 1. Add Spec.Installation.InstallConfigSecret containing a full openshift-install InstallConfig  to pass directly to installer.
+   * This will allow immediate use of new openshift-install features without needing to mirror in Hive's API.
+   * Ideally most items in ClusterDeployment.Spec should be assumed mutable, however this will not universally be true.
+   * There will be some overlap, information that must be provided in both the InstallConfig and the ClusterDeployment. We will attempt to error if you mismatch this information whenever possible.
+ 1. Add Spec.Installation.ManifestsConfigMap for injecting additional manifests into the install process.
+   * Can be used for something like switching to Calico for networking.
+   * ConfigMap with a file in each key. Will mount the whole configmap and iterate each file, adding to the installer's manifests dir.
+ 1. Remove:
+   * Spec.ClusterName - In install config, not relevant thereafter.
+     * May be used in some naming, but hopefully can transition to CD.Name instead.
+   * Spec.ControlPlane - Not relevant after install today as masters are not a MachineSet. Can be re-added in future if necessary.
+   * Spec.Networking
+ 1. Remaining immutable ClusterDeployment.Spec fields:
+   * Spec.ImageSet and Spec.Images. (technically could be mutable but has no effect once we've installed)
+   * Spec.Platform.AWS.Region (deemed important enough to be present in spec)
+   * ManageDNS bool: Enabled automated DNS forwarding. May be removed in V2.
+   * BaseDomain: Used to drive the automatic creation of the DNSZone if manageDNS is true. May be removed in V2.
+ 1. Introduce Spec.ClusterMetadata struct with fields for InfraID, ClusterUUID, and admin password/kubeconfig secret references.
+   * All of these fields will be removed from Status where they cannot be restored.
+   * InfraID and ClusterUUID will be programatically mutable, but once set and the cluster is marked installed, validation will make them immutable.
+   * ClusterMetadata may be renamed if a better idea comes up.
+ 1. Rename ClusterDeployment.Spec.SSHKey to ClusterDeployment.Spec.SSHKeySecret to better match all other secret references.
+   * Should not be removed as private keys are not contained in InstallConfig and are used by Hive.
+   * Rename to SSHPrivateKeySecret and should only contain private key going forward.
+ 1. Rename CertificateBundle.SecretRef to CertificateSecret.
+ 1. Move PlatformSecrets into Platform, rename Credentials to CredentialsSecret.
+ 1. Add support for bare metal provisioning.
+   * At this point, we do not see any major implications for BareMetal. This is modelled as a separate platform in install config (see [this example](https://github.com/openshift/installer/blob/master/docs/user/metal/install_ipi.md#install-config))
+   * This would be specified in the InstallConfig.
+   * Hive would likely want a platform indicator explicitly for bare metal if we need a special invocation of the installer.
```

Comment:
The installer binary for baremetal has a different name because it is compiled with libvirt support (where the normal openshift-install binary is not).

Reply:
This is a good point, we would need that installer. Presumably Hive is extracting the installer from the release image? With `oc`, you're able to do this on 4.2 and 4.3:

`oc adm release extract --command='openshift-baremetal-install'  <release image>`

It's a hidden command that's extractable (not listed in help).

Reply:
Interesting, ok that's good to know, we do extract the openshift-install command in similar fashion, we'll need to do the same for this one if it's a bare metal cluster.

Reply:
@stbenjam does the '<release image>' argument there refer to a file, URL, or something else? We're going to need to be able to do this disconnected, with a local image registry other than quay.io.

Reply:
That's a good point, it is a pull spec: registry.svc.ci.openshift.org/ocp/release:4.3.0-0.ci-2019-11-04-170858.

Presumably we could pass the real image pullspec in Artifactory instead, but how does Hive handle the disconnected case? I have not seen this in reality yet but I were told that the installer can dynamically swap out quay.io URL's for a local registry for the disconnected case.  We'll probably need other components to handle this, too, like Hive.

Reply:
Presumably oc uses the pull spec to build a URL to download something, so if we can give it a different URL it can use that instead.

Reply:
Sure, seems like the ClusterDeployment resource specifies the cli and installer images rather than the release pullspec, https://github.com/openshift/hive/pull/631/files#r342248335

I suppose we would need to give it local URL's for those.

Reply:
Those are just status fields set by extracting from release image. 

I'm not sure how the installer dynamically swaps registries, but if they've got something we can definitely expose that.